src/qibo/quantum_info/quantum_networks.py

```diff
@@ -629,53 +565,497 @@ def _run_checks(self, partition, system_output, pure):
-                self.system_output[k * 2] = False
-            self.system_output = tuple(self.system_output)
+        super().__init__(
+            tensor, partition, [True, False] * (len(partition) // 2), pure, backend
+        )
+
+    def is_causal(
+        self, order: Optional[Union[int, str]] = None, precision_tol: float = 1e-8
+    ):
+        """Returns bool indicating if the Choi operator :math:`\\mathcal{J}` of the network satisfies the causal order condition.
+
+        Causality is calculated based on a recursive constrains.
+        This method reduce a n-comb to a (n-1)-comb at each step, and checks if the reduced comb is independent on the
+        last output system.
+
+        Args:
+            order (str or int, optional): order of the norm. Defaults to ``None``.
+            precision_tol (float, optional): threshold :math:`\\epsilon` that defines
+                if Choi operator of the network is :math:`\\epsilon`-close to causality
+                in the norm given by ``order``. Defaults to :math:`10^{-8}`.
+
+        Returns:
+            bool: Causal order condition.
+        """
+        if precision_tol < 0.0:
+            raise_error(
+                ValueError,
+                f"``precision_tol`` must be non-negative float, but it is {precision_tol}",
+            )
+
+        if order is None and self._backend.__class__.__name__ == "TensorflowBackend":
+            order = "euclidean"
+
+        backend = self._backend
+
+        dim_out = self.partition[-1]
+        dim_in = self.partition[-2]
+
+        trace_out = backend.cast(
+            trace(dim_out, backend=backend).full(), dtype=self._tensor.dtype
+        )
+        trace_in = backend.cast(
+            trace(dim_in, backend=backend).full(), dtype=self._tensor.dtype
+        )
+
+        if self._backend.__class__.__name__ == "PyTorchBackend":
+            reduced = self._tensordot(self.full(), trace_out, dims=([-1], [0]))
+            sub_comb = self._tensordot(reduced, trace_in, dims=([-1], [0]))
+            expected = self._tensordot(sub_comb, trace_in / dim_in, dims=0)
         else:
-            self.system_output = tuple(self.system_output)
+            reduced = self._tensordot(self.full(), trace_out, axes=(-1, 0))
+            sub_comb = self._tensordot(reduced, trace_in, axes=(-1, 0))
+            expected = self._tensordot(sub_comb, trace_in / dim_in, axes=0)
+
+        norm = self._backend.calculate_norm(reduced - expected, order=order)
 
-    def _full(self):
-        """Reshapes input matrix based on purity."""
-        matrix = self._backend.cast(self._matrix, copy=True)
+        if float(norm) > precision_tol:
+            return False
+        elif len(self.partition) == 2:
+            return True
+        else:
+            return QuantumComb(
+                sub_comb, self.partition[:-2], pure=False, backend=self._backend
+            ).is_causal(order, precision_tol)
+
+    @classmethod
+    def from_nparray(
```

Comment:
Since the class accepts arrays of all backends, the method should be renamed to something more agnostic like `from_array` (or `from_tensor` or another suggestion).

Reply:
I haven't tested the constructor from other data types. I'm not sure if this function can construct `QuantumNetwork` from all data types properly. I remember I used some `numpy` function in the intermediate steps. We may need more tests if we want it to be universal.

Reply:
Excluding something out of the norm, everything in the `quantum_info` module (and `qibo` in general) should be backend agnostic.